File: Go/BinarySearch.go
package main

import (
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)


func main() {
	var array []int
	var elementToFind int

	// Converts all input arguments to integers and adds them to the list to be sorted
	for _, value := range os.Args[1:] {
		number, err := strconv.Atoi(value)

		if err != nil {
			log.Fatal(err)
		}

		array = append(array, number)
	}

	// Sort array of ints in increasing order
	sort.Ints(array)

	// Asks the user to enter the number they want to find within the array
	fmt.Print("Enter the number to find: ")
	fmt.Scanf("%d", &elementToFind)

	fmt.Println(fmt.Sprintf("Element %d was found at position %d", elementToFind, search(array, elementToFind)))
}

/*
I => Array: 12 19 20 33 44 68 79 80
	 Element to find: 68
O => 5

Time complexity: O(log n)
Space complexity: O(1)
*/

File: Python/Binary_Search.py
# Binary Search = An efficent way to search in a sorted list
def binary_new_function(arr, low, high, x):
    print("NEW function in binary search")
    while low < high:
        mid = low + (high - 1) // 2
        if (arr[mid] == x):
            return mid
        elif (arr[mid] < x):
            high = mid - 1
        else:
            low = mid + 1
    return -1


def binary(arr, low, high, x):
    while low < high:
        mid = low + (high - 1) // 2
        if (arr[mid] == x):
            return mid
        elif (arr[mid] < x):
            low = mid + 1
        else:
            high = mid - 1
    return -1


print('Enter a count of number in a list or array')
arr = list(map(int, input().split()))
n = len(arr)

arr.sort()
x = int(input("Enter the number to find\n"))
result = binary(arr, 0, len(arr) - 1, x)
if (result == -1):
    print("Element not found")
else:
    print("Element found at ", result + 1, " position")

'''
Time complexity ==> O(logn)
Space Complexity ==> O(1)

I/O ==>

Enter a list of number
234 567 321 345 123 110 767 111
Enter the number to find
767
Element found at  8  position
'''
